### ISSUE TITLE
PyTorch loss functions type cast to the first argument, rather than  type promoting all arguments.

### ISSUE THREAD

**Issue by Chillee** [pytorch/pytorch#76870]
### 🐛 Describe the bug

@mruberry suggested this might be a bug, so filing it (https://github.com/pytorch/pytorch/pull/76633#issuecomment-1117180656).

```
import torch
import torch.nn.functional as F

print(F.l1_loss(torch.randn(5, dtype=torch.half), torch.randn(5, dtype=torch.float64)).dtype)
```
prints `torch.float16`.

cc: @mruberry , @ezyang , @ngimel 

### Versions

Nightly

cc @nairbv @mruberry @ezyang @ngimel

**Comment by gchanan** [comment]
My guess is this is accidental, or at least not thought through.  Basically at some point we rewrote l1_loss to not just be the simple math formula because folks objected to the broadcasting.  Later we put in a warning about the broadcasting, but probably never thought about the type promotion.

**Comment by khushi-411** [comment]
Hi, @Chillee!
I was trying to reproduce this, looks like it is resolved (branch: `2.0.0a0+gita2ac25f`). Just thought to ping :) Thanks!
```python
>>> import torch
>>> import torch.nn.functional as F
>>> print(F.l1_loss(torch.randn(5, dtype=torch.half), torch.randn(5, dtype=torch.float64)).dtype)
torch.float64
```
